### PR TITLE
Make the gui easier to modify by splitting the main widget in multiple sub-widgets

### PR DESCRIPTION
**Pull request by fdrgsp**
Instead of having a single `.ui` file for the whole `napari-micromanager` gui, now there are multiple sub-widgets in the `_gui_objects` folder.

`.ui` files are used only for the more complex `mm_xyz_stage.ui`, `multid_gui.ui`, and `explore_sample.ui`.

The gui is now created in `_gui.py`.

In this PR I have also to the gui the comboboxes to allow to select multiple stage devices (as suggested in #29) (for now only XY and Z. Offset/Autofocus devices not implemented yet).

Screen Shot 2022-02-21 at 6 27 40 PM

### CODE REVIEW

**Comment by ianhi** [comment]
@fdrgsp I'm happy to review this, but if you could do the below two things it would make it much easier to do so:

1. Mark as draft until you think it is done (is it done already?)
2. Do a rebase and squash your commits into a few commits where each commit is one logical change. This would make it much easier to review.
You can see how to squash commits together here: https://thoughtbot.com/blog/git-interactive-rebase-squash-amend-rewriting-history#squash-commits-together

**Comment by tlambert03** [comment]
I think this is definitely a step in the right direction!  I'd be inclined to merge this sooner than later to avoid conflicts with other PRs.  @fdrgsp, can you comment on whether you were able to resist the urge to modify _behavior_ for the most part?  All the file changes naturally make it hard to compare the diffs.  I think this is what @ianhi was referring to.  But, if you tried hard to restrict this _just_ to UI/layout changes, with no functionality changes... and if all the tests are passing (which they appear to be, besides windows), then I think I'd be ok generally getting this in and following up with smaller cleanups afterwards.  Can you think of anything you might have touched that wasn't strictly UI related?

One thing that I think is changing here is layout padding.  So the UI is starting to take up more empty space.  You might want to try `setContentsMargins(0,0,0,0)` on some of the new layout objects to avoid all that dead space.

Sidenote (_not_ to be added to this PR!) I think we should either start letting some of these subwidgets be their own thing,  such that they can be added to the window and popped out independently of the main widget.   Alternatively, we could start using QCollapsible from superqt to save space (as shown below):

https://user-images.githubusercontent.com/1609449/154854854-5cf46307-a478-427c-80cb-651d183e4659.mov

but let's definitely not do that here.

**Comment by fdrgsp** [comment]
@tlambert03 

> Can you think of anything you might have touched that wasn't strictly UI related?

No I don't think I modified any behavior. The only think I have changed a little is in `multid_widget.py` and `explore_sample.py`. 
- In `multid_widget.py` I added the possibility to run the experiment even if there is not a motorized xy and/or z stage (see `_get_state_dict()`).

- For `explore_sample.py`, in `main_window.py` I disabled the explorer tab if there is no xy stage.

> One thing that I think is changing here is layout padding. So the UI is starting to take up more empty space. You might want to try setContentsMargins(0,0,0,0) on some of the new layout objects to avoid all that dead space.

I'll work on that!

**Comment by ianhi** [comment]
> Sidenote (not to be added to this PR!) I think we should either start letting some of these subwidgets be their own thing, such that they can be added to the window and popped out independently of the main widget. Alternatively, we could start using QCollapsible from superqt to save space (as shown below):

wow yes please! yes to both probably, but QCollapsible as a start that can (per above gif) be implemented quickly

**Comment by fdrgsp** [comment]
> One thing that I think is changing here is layout padding. So the UI is starting to take up more empty space. You might want to try setContentsMargins(0,0,0,0) on some of the new layout objects to avoid all that dead space.

@tlambert03 I think it looks better now

**Comment by tlambert03** [comment]
@ianhi, lemme know if you want to peek at this one.  Will likely merge tomorrow (this one is very prone to merge conflicts)

File: micromanager_gui/main_window.py
Comment (ianhi) [inline micromanager_gui/main_window.py:129]: I think it makes more sense to move these signal connections (and respective callback methods) into the sub widgets. It feels a bit awkward to have the main window be responsible for managing the state of the sub widgets.

Also this would probably force us to be better about using the `mmc` object as the real source of truth (sorta the model of MV framework) rather than having that work leak into the main window.

File: micromanager_gui/main_window.py
Comment (ianhi) [inline micromanager_gui/main_window.py:237]: is there any changes in this whole section? I'm not sure why this is showing up on github, but a lot of this I think is showing up due to how you pulled in new changes (as commits rather than merge commits or rebases)

File: micromanager_gui/_gui_objects/_mm_configuration_widget.py
Comment (ianhi) [inline micromanager_gui/_gui_objects/_mm_configuration_widget.py:32]:
```suggestion
        self.cfg_LineEdit = QtW.QLineEdit()
        self.cfg_LineEdit.setPlaceHolderText("MMConfig_demo.cfg")
```

Comment (ianhi) [reply micromanager_gui/_gui_objects/_mm_configuration_widget.py:32]: this placeholder got lost in the move away from `ui` file.

File: micromanager_gui/_gui_objects/_camera_widget.py
Comment (ianhi) [inline micromanager_gui/_gui_objects/_camera_widget.py:61]: For me at least this is not rendering correctly:
image

Comment (ianhi) [reply micromanager_gui/_gui_objects/_camera_widget.py:61]: turns out this was bad before for me. Not a blocker

**Review by ianhi — CHANGES_REQUESTED** [review]
Overall I'm a big fan of this! I think it will it much easier to get a handle on what's happening for any new contributors and also makes it easier to make changes and improvements!

That said I'd suggest holding off a little bit for two reasons:

1. (Simple) fix placeholder text in config lineedit - see comment
2. Moving these around is good time to consider which widgets are responsible for managing what state and where signals should be connected. 


For 2 I'd suggest that each of the subwidgets is responsible for managing the relevant core signals.  And each subwidget should provide a `refresh` method that the main window can call when a new config. 

not 100% on this, but think that now is a good time to make a decision as to where signal connections should be happening.

File: micromanager_gui/_gui_objects/_camera_widget.py
Comment (fdrgsp) [reply micromanager_gui/_gui_objects/_camera_widget.py:61]: As in #107 I think we can remove the widgets for binning and bit depth.

File: micromanager_gui/_gui_objects/_mm_configuration_widget.py
Comment (fdrgsp) [reply micromanager_gui/_gui_objects/_mm_configuration_widget.py:32]: I'll put it back in the widget.

File: micromanager_gui/main_window.py
Comment (fdrgsp) [reply micromanager_gui/main_window.py:237]: I think this is just because I reordered the methods

File: micromanager_gui/main_window.py
Comment (fdrgsp) [reply micromanager_gui/main_window.py:129]: @ianhi  I think this can be a good idea! However it is maybe better to merge this "new" way to make the gui before doing any further changes...otherwise it might be harder to deal with conflicts. In addition, once this PR is merged, it will be quite easy to merge #107 which can be the new starting point to change widgets signals, etc.... why do you think?

**Comment by tlambert03** [comment]
> That said I'd suggest holding off a little bit for two reasons:
> 2. Moving these around is good time to consider which widgets are responsible for managing what state and where signals should be connected.
> For 2 I'd suggest that each of the subwidgets is responsible for managing the relevant core signals. And each subwidget should provide a `refresh` method that the main window can call when a new config.
> not 100% on this, but think that now is a good time to make a decision as to where signal connections should be happening.


In general, I agree with all these comments.  I definitely agree that in general each subwidget should be more self-contained, and main window should simply compose them all.  Signals should be contained on subwidgets as much as possible, and state should be made clearer (as in #109).  I don't, however, think that this PR is a good time to change all that.  It's not a new problem, but having this PR in would make it easier to reason about all that stuff and fix in smaller followup PRs.  Meanwhile, as long as this stays open, pretty much every other PR is going to have to either wait or cause a conflict that requires manual resolution.

If we generally feel like this is a good direction, i think we clean up the little remaining issues and merge... and then (**!**) lets get back to much smaller PRs that try to solve a single problem.  @fdrgsp, lemme know when you're done making changes and I'll do a final thorough review

**Comment by fdrgsp** [comment]
>  lemme know when you're done making changes and I'll do a final thorough review

@tlambert03 I think you can review it!

**Comment by tlambert03** [comment]
two quick observations you can work on while I continue to look:

1) why does `MainWindow` call `self.create_gui` instead of `MicroManagerWidget` ... when it seems to be entirely contained to `MicroManagerWidget`?  

2) something has been lost with regards to enabling the gui.  On main, if you run `python launch-dev.py`, everything is active and ready to go.  Here, it appears that one _must_ interact with the GUI (and click the load button) to make everything active.  We definitely have to preserve the "programmatic" activation as well.  The gui should be smart enough to know if it's ready to go, without expecting a specific manual action in the GUI

File: tests/conftest.py
Comment (fdrgsp) [reply tests/conftest.py:28]: @tlambert03 

Something like this?

```
...
sig.systemConfigurationLoaded.connect(self._on_system_cfg_loaded)
...
def _on_system_cfg_loaded(self):
        if len(self._mmc.getLoadedDevices()) > 1:
            self._set_enabled(True)
```

Comment (tlambert03) [reply tests/conftest.py:28]: sure, if that works!

**Comment by ianhi** [comment]
> something has been lost with regards to enabling the gui. On main, if you run python launch-dev.py, everything is active and ready to go. Here, it appears that one must interact with the GUI (and click the load button) to make everything active. We definitely have to preserve the "programmatic" activation as well. The gui should be smart enough to know if it's ready to go, without expecting a specific manual action in the GUI

On main this happens due to these lines:
https://github.com/tlambert03/napari-micromanager/blob/eac0da7c33671ff7210e1f6e9ec5a9446329a457/micromanager_gui/main_window.py#L183-L184

> but having this PR in would make it easier to reason about all that stuff and fix in smaller followup PRs. Meanwhile, as long as this stays open, pretty much every other PR is going to have to either wait or cause a conflict that requires manual resolution.

> If we generally feel like this is a good direction, i think we clean up the little remaining issues and merge... and then (!) 


Yup sounds good! My only worry here is that there are some other small things like the programmatic launching that may have been lost bc we don't have complete tests and the diff is so big. But short of breaking this up into a series of much smaller PRs I don't think there much of a way around that. So modulo fixing the two issues talley noted :+1:  from me.

**Comment by tlambert03** [comment]
just pushed some changes @fdrgsp.  Couple things to point out:

1) i saw a lot of QGridLayout usage when all that was needed was a QVBox/QHBoxLayout.  In general, if you don't _need_ a grid, use one of the simpler layouts instead
2) i noticed a number of attributes being named one thing in the MicroManagerWidget (eg. `self.mm_configuration`) only to be renamed (eg. `self.cfg`) in the MainWindow subclass.  Try not to rename things gratuitously as it makes it harder to get a sense for what a class really _is_ (for example, now there's a mental burden to remember that those two things are actually just two names for the same thing)

If tests pass, I will probably merge... since I think this is heading in the right direction and I'm rather scared of merge conflicts now.  After this merges, I'll merge it into #107  and merge that too.  

_then_.... I'd really like to enforce #111.  I'd like some time to go through all the code, minimize duplication and simplify code, and shore up our tests.  I agree with @ianhi's statement that it's scary to make changes like this without having great test coverage.  so let's get that done before moving forward.

thanks for this big effort @fdrgsp!

**Pull request merged**
